src/main/java/com/couchbase/lite/Database.java

```diff
@@ -1152,27 +1149,54 @@ private void markBlobstoreMigrated() {
      */
     @InterfaceAudience.Private
     public boolean close() {
-        if(!open) {
+        if (!open) {
             return false;
         }
 
-        if(views != null) {
+        if (views != null) {
             for (View view : views.values()) {
                 view.databaseClosing();
             }
         }
         views = null;
 
-        if(activeReplicators != null) {
-            for(Replication replicator : activeReplicators) {
+        if (activeReplicators != null) {
+            List<CountDownLatch> latches = new ArrayList<CountDownLatch>();
+            for (Replication replicator : activeReplicators) {
+                // handler to check if the replicator stopped
+                final CountDownLatch latch = new CountDownLatch(1);
+                replicator.addChangeListener(new Replication.ChangeListener() {
+                    @Override
+                    public void changed(Replication.ChangeEvent event) {
+                        if (event.getSource().getStatus() == Replication.ReplicationStatus.REPLICATION_STOPPED) {
+                            latch.countDown();
+                        }
+                    }
+                });
+                latches.add(latch);
+
+                // ask replicator to stop
                 replicator.databaseClosing();
             }
+
```

Comment:
Response for question 2:
- If user calls Database.close() from UI thread, it might cause UI freezing state for 20 sec maximum. SQLiteDatabase.close() does not block, it just release references.
- However, Database.close() is annotated with `@InterfaceAudience.Private`. So user should not call this directly.
- We could provide non-blocking Database close() method if necessary.